### DISCUSSION TITLE
Issue in file upload using graphql-upload

### DISCUSSION THREAD

**Discussion by payal8797** [apollographql/apollo-server#7892] [Q&A]
I want to implement csv upload ( timestamp: String, activePower: Float, energy: Float) using nodejs, apollo graphql.

Implementation:

**typeDefs.js**
```javascript
const { gql } = require('apollo-server');
module.exports = gql`
scalar Upload
type File {
    filename: String!
    mimetype: String!
    encoding: String!
}
type Query {
    getCSVData: [File!]
}
type Mutation {
    uploadCSVFile(file: Upload!): File
}
`
``` 

**model**:

```javascript
const mongoose = require('mongoose');
const csvDataSchema = new mongoose.Schema({
    timestamp: {type: Date, required: true  },
    activePower: { type: Number, required: true},
    energy: { type: Number,required: true }
});
module.exports = mongoose.model('CSVData', csvDataSchema);
``` 

**index.js**

```javascript

const { ApolloServer } = require('apollo-server-express');
const express = require('express');
const {graphqlUploadExpress} = require('graphql-upload');
const mongoose = require('mongoose');
const typeDefs = require('./graphql/typeDefs');
const resolvers = require('./graphql/resolvers');
const MONGODB = "mongodb://localhost:27017";
const startServer = async () => {
  const app = express();
  app.use(graphqlUploadExpress({ maxFileSize: 10000000, maxFiles: 10 }));
  const server = new ApolloServer({
    typeDefs,
    resolvers,
  });
  await server.start();
  server.applyMiddleware({ app });
  mongoose.connect(MONGODB, { useNewUrlParser: true, useUnifiedTopology: true })
    .then(() => {
      console.log("MongoDB Connected");
      app.listen({ port: 5000 }, () => {
        console.log(`Server running at http://localhost:5000${server.graphqlPath}`);
      });
    });
};
startServer();

``` 

resolver: CSVData.js
```javascript
const {GraphQLUpload}  = require('graphql-upload');
const CSVData = require('../../models/CSVData');
const parse = require('csv-parse');
module.exports = {
  Upload: GraphQLUpload,
  Mutation: {
    async uploadCSVFile(_, {file}) {
      const { createReadStream, filename, mimetype, encoding } = await file;
      console.log("file", file, createReadStream, filename, mimetype, encoding)
      const stream = createReadStream();
      return new Promise((resolve, reject) => {
        const parser = parse({
          columns: true,
          skip_empty_lines: true
        });

        stream.pipe(parser);

        parser.on('data', async (row) => {
          const { timestamp, activePower, energy } = row;

          const newData = new CSVData({
            timestamp: new Date(timestamp),
            activePower: parseFloat(activePower),
            energy: parseFloat(energy)
          });

          await newData.save();
        });

        parser.on('end', () => {
          resolve(`File ${filename} uploaded successfully`);
        });

        parser.on('error', (err) => {
          reject(`Error parsing CSV: ${err.message}`);
        });
      });
    }
  },
  Query: {
    async getCSVData() {
      try {
        const data = await CSVData.find();
        return data;
      } catch (err) {
        throw new Error(err);
      }
    }
  }
};

``` 

packages:
"apollo-server": "^3.5.0",
"apollo-server-errors": "^3.3.1",
"apollo-server-express": "^3.13.0",
"bcryptjs": "^2.4.3",
"csv-parse": "^5.5.6",
"csv-parser": "^3.0.0",
"express": "^4.19.2",
"graphql": "^16.8.2",
"graphql-upload": "^13.0.0",
"jsonwebtoken": "^9.0.2",
"mongoose": "^6.13.0",
"multer": "^1.4.5-lts.1",
"nodemon": "^2.0.22"

Issue:
1. When I hit the api using postman and console "createReadStream, filename, mimetype, encoding": all of them are undefined. 
2. I get an error: createReadStream is not a function. To avoid this , I replaced
         const { createReadStream, filename, mimetype, encoding } = await file;
         with
        const { stream, filename, mimetype, encoding } = await file;

3. Then I get another error: parse is not a function
4. My graphql-upload version is 13.0.0. I am not able to use latest version of graphql-upload( https://github.com/jaydenseric/graphql-upload/issues/305#issuecomment-1136574019) . I tried following https://stackoverflow.com/questions/72361047/error-no-exports-main-defined-in-graphql-upload-package-json but had no success. I think it works woth Typescript and I am using Javascript.

I am not able to understand the reason for any of the above issues, I am following the official documentation but not able to understand where exactly is the problem.

Any help would be greatly appreciated. :)

**Comment by Bilal798** [comment]
I'm also facing the same problem. :(

**Comment by glasser** [comment]
This looks like an issue with `graphql-upload`! You should report it over there. We did have a built-in integration with graphql-upload way back in Apollo Server 2 (which has been end-of-life since October) but we removed that. It also looks like you're trying to use Apollo Server 3 (or older) via `apollo-server-express` rather than `@apollo/server`; AS3 is deprecated and will itself be end-of-life this upcoming October.
